windows-msi/boot.wsf.in

```diff
@@ -0,0 +1,112 @@
+<?xml version="1.0"?>
+<!--
+  openvpn-build — OpenVPN packaging
+
+  Copyright (C) 2018 Simon Rozman <simon@rozman.si>
+
+  This program is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License version 2
+  as published by the Free Software Foundation.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License along
+  with this program; if not, write to the Free Software Foundation, Inc.,
+  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+-->
+<package>
+	<job>
+		<runtime>
+			<usage>Lauches architecture-specific @PRODUCT_NAME@ MSI package for install.
+
+@PRODUCT_NAME@-install-@PRODUCT_VERSION@.exe [Additional msiexec.exe parameters]
+
+
+Notes:
+
+- This script launches msiexec.exe /i on selected MSI package.
+- Logging /l* to %TEMP%\@PRODUCT_NAME@-install.log file is arranged.
+- All command line parameters passed to this script are forwarded to
+  msiexec.exe call:
+  - See msiexec.exe /? for reference.
+  - Do not pass /i or /l as those are already used by this script.
+
+
+Examples:
+
+@PRODUCT_NAME@-install-@PRODUCT_VERSION@.exe ADDLOCAL=driver,util,sdk
+   Install "driver", "util" and "sdk" features.
+
+@PRODUCT_NAME@-install-@PRODUCT_VERSION@.exe INSTALLDIR=&lt;folder&gt;
```

Comment:
Will @PRODUCT_VERSION@ include the package version? For example, with NSIS, we have a version string like 2.4.6-I601, where I601 represents the package version. We increment the package version whenever dependencies (mainly openssl) need to be updated.

Reply:
Unfortunately, no. But, I can add that to the filename.

For package versioning with MSI we shall use:
- same product version again
- same product code GUID
- change the package code GUID (which is auto-generated by WiX, therefore all we need is to rebuild the MSI packages)

Source: https://docs.microsoft.com/en-us/windows/desktop/Msi/productversion

I have added a bullet on my TODO list to test MSI small updates (same product, different package) with OpenVPN packages explicitly.

Reply:
> "Houston, we have a problem..."

I have introduced the package version support and played with it for a while.

1. MSI will refuse to install a .msi package with the same product version, product code, and different package code over the existing one. MSI considers this a _small update_ and only way to install it is to create and deploy a .msp patch. This will complicate our deployment flow: you'll need to publish all .msp patches for Group Policy admins, I'll need to add .msp patches to SFX .exe installer payload and update the bootstrapper appropriately...

2. Next option is so-called _minor update_: we keep the product code but change the product version. An example of a minor update is a service pack. We can deploy it as a .msp patch or a full-fledged .msi package. The later option suits us best. But (I hate this word), the MSI honors only first three version fields. "2.4.6" or "2.4.6.1" is all the same product version for the MSI. Which won't fly with existing OpenVPN versioning scheme. One possible workaround is we shift version left by one. But only for MSI's `ProductVersion` property. Filenames and MSI GUI would still display "OpenVPN 2.4.6-1". Imagine shifting "openvpn-2.4.6-1" to "openvpn2-4.6.1". It is still possible to author OpenVPN3 MSI as an upgrade to OpenVPN2 MSI even with this change.
The only negative side effect is the legacy Programs and Features app from Control Panel will display product version as 4.6.1 in the package info.

I suppose we can't easily just quickly release 2.4.7 should some dependency require an immediate CVE deployment, right?

Reply:
Instead of shifting off the major version (2), what about combining release+build: say 2.4.6.0 is treated as 2.4.60, 2.4.6.2 as 2.4.62 etc. As long as we wont have more than 10 (or 16 if converted as hex) "minor" updates, this would work. Not ideal but better than version becoming 4.6.1 etc.

Reply:
We've never had more than 3-4 NSIS installer releases. So ten is more than adequate. It will will be confusing, but probably less confusing than having OpenVPN 4.6.1 show up on the system :smile:.

Reply:
Selva, that's very clever, thanks - I shall go for it. Samuli, the product version as authored by `ProductVersion` property in MSI is not displayed a lot.

Quoting the MSDN:

>
>
> The first field is the major version and has a maximum value of 255. The second field is the minor version and has a maximum value of 255. The third field is called the build version or the update version and has a maximum value of 65,535.

This means we have plenty of space in the third version field. So, let's use a factor of 100:
2.4.600, 2.4.601 etc.

Reply:
Resolved in db3deac24002e327a474e458ed9f1dfdbb25788e. However, I still need to thoroughly test minor updates before going live.